### PR TITLE
Database Compression 

### PR DESCRIPTION
**Pull request by rohit-cmu**
## Files Added

* [compressed_tile.cpp](https://github.com/rohit-cmu/compression-peloton/blob/master/src/storage/compressed_tile.cpp)
* [compressed_tile.h](https://github.com/rohit-cmu/compression-peloton/blob/master/src/include/storage/compressed_tile.h)

### Other Existing Files Modified

Some additional functions have been added to support database compression. Here are the following files:

* [tile.h](https://github.com/rohit-cmu/compression-peloton/blob/master/src/include/storage/tile.h)
* [tile.cpp](https://github.com/rohit-cmu/compression-peloton/blob/master/src/storage/tile.cpp)
* [tile_group.h](https://github.com/rohit-cmu/compression-peloton/blob/master/src/include/storage/tile_group.h)
* [tile_group.cpp](https://github.com/rohit-cmu/compression-peloton/blob/master/src/storage/tile_group.cpp)
* [data_table.cpp](https://github.com/rohit-cmu/compression-peloton/blob/master/src/storage/data_table.cpp)

## Strategy for Compressing

* Once a tile gets full (all slots occupied by tuples), the CompressTile is created.
* We scan each column and sort it.
* We compute the median, min and max, this gives us the minoffset(min-median) and maxoffset(max-median).
* If these offsets can be represented in a smaller data type than the original data type, we compress the entire column.
* The median is chosen as the base, and essentially only the offsets are stored in the column.
* Currently we can compress SMALLINT and INTEGER, BIGINT, DECIMAL. TINYINT is already 1 byte so we do not compress it.
* We also support deduplication for VARCHAR field
* We are yet to add support for decimal values.
* This median is also stored as the metadata to later retrieve the original value

## Testing

* [Compression Correctness Test](https://github.com/rohit-cmu/compression-peloton/blob/master/test/sql/compression_sql_test.cpp)
* [Compression Size Test](https://github.com/rohit-cmu/compression-peloton/blob/master/test/storage/compression_test.cpp)

### Compression Correctness Test:
* This test inserts 25 tuples
* Each tuple is of the form (i, i*100) where i belongs to (0,25)
* Since each tile group contains 1 tile and 10 tuples per tile, there are 3 tile groups formed.
* The first 2 tile groups are full and get compressed.
* The third tile group has 5 slots vacant and is still not full and is uncompressed.
* Thus we now have compressed and uncompressed data.
* We now perform a SELECT * on this and expect to correctly recieve the true value of the compressed data and uncompressed data.

### Compression Size Test:
* This test inserts 100 tuples
* Each tuple is of the form [Integer, Integer, Decimal, Varchar]
* Thus the tuple length is (4+ 4+ 8+ 8) = 24 bytes.
* We currently support the compression of integers
* The integers get compressed to TINYINT making the tuple sizes now (1+ 1+ 8+ 8) = 18 bytes
* The test checks this decrease in size.

### CODE REVIEW

**Comment by vanryan** [comment]
I feel it is not a very good practice to include irrelevant (other people's) commits and changed files into one's own pull request... It takes a lot of energy to pick out theme-related files to review and merge conflict might occur if being merged into master...

File: .gitignore [hunk 1]
```diff
@@ -313,3 +313,6 @@ third_party/gflags/
 
 # Test output file logs to ignore
 stats_log
+
+test.txt
```
Comment (vanryan) [inline .gitignore]: This should not be included into the pull request to the master, right... cuz it is for own specific testing use.

File: .travis.yml [hunk 2]
```diff
@@ -16,6 +17,7 @@ matrix:
           addons:
               apt:
                 sources:
+                    - llvm-toolchain-precise-3.7
```
Comment (vanryan) [inline .travis.yml]: I guess this change was from llvm codegen

File: CodeReview#1_README.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
+## Files Added
```
Comment (vanryan) [inline CodeReview#1_README.md]: Why is this file included in this pull request...

**Comment by wangxiaoyou11** [comment]
I'm surprised by your PR size with 300+ commits. It's really hard to review your PR because there are so many duplicate commits with `cmu-db:master`. I can hardly find what you have changed among those 272 files.
I think you should `rebase` the `cmu-db:master` instead of doing a `merge`. Could you please make this PR clean first?

**Comment by rohit-cmu** [comment]
@vanryan @wangxiaoyou11 We are fixing this. New Pull request should be available soon!

**Comment by poojanilangekar** [comment]
@vanryan @wangxiaoyou11 

Link to the new Pull Request : https://github.com/cmu-db/peloton/pull/668

Sorry for the inconvenience!
